src/Workspaces/Core/Portable/Workspace/Solution/Document.cs

```diff
@@ -388,13 +388,35 @@ public Document WithSourceCodeKind(SourceCodeKind kind)
     /// Creates a new instance of this document updated to have the text specified.
     /// </summary>
     public Document WithText(SourceText text)
-        => this.Project.Solution.WithDocumentText(this.Id, text, PreservationMode.PreserveIdentity).GetRequiredDocument(Id);
+    {
+        var solution = this.Project.Solution.WithDocumentText(this.Id, text, PreservationMode.PreserveIdentity);
+
+        if (Id.IsSourceGenerated)
+        {
+            var document = solution.GetRequiredProject(Id.ProjectId).TryGetSourceGeneratedDocumentForAlreadyGeneratedId(Id);
+            Contract.ThrowIfNull(document, "We just modified the text of the generated document, so it should be available synchronously");
```

Comment:
yes.  i like these asserts.  stating that this would only be ok if it must have been generated.  as much as possible i want that.   or even just having a GetSourceGeneratedDocumentForAlreadyGeneratedId, which docs that it *will* throw and that the caller must only be cases that know that this must have been the case.

Reply:
I'm updating this and other spots to call `GetSourceGeneratedDocumentForAlreadyGeneratedId`, and adding more docs to that.